Integration of fast simulation module in O2 pipeline

SwirtaB opened the pull request:
Integration allows to to run fast simulation model on event. FastSim module is configured via parameters.

SwirtaB committed:
Added parameters for fastsim integration

SwirtaB committed:
Integrate fastsimulation into ZDC Detector

Modified CMake to link against FastSimulation lib
Modified Detector class:
- Added pointer to FastSimulation object as member
- Added container for results as member
- Added required changed to Detector's constructor

SwirtaB committed:
Added fastsim dispatch

Dispatch produce empty result for now
Fixed compile guard

SwirtaB committed:
Added additional functions to interface

Modified run example to work with changed interface

SwirtaB committed:
Fixed typo

SwirtaB committed:
Dispatch integrated

SwirtaB committed:
Refactored integration

FastSim guards are more robust
Static_casts for input data
Deleted obsolete prints

SwirtaB committed:
Fixed o2checkcode error

sawenzel commented on Detectors/ZDC/simulation/src/Detector.cxx:
```diff
@@ -2398,6 +2430,22 @@ void Detector::BeginPrimary()
   mCurrentPrincipalParticle = *stack->GetCurrentTrack();
+
+#ifdef ZDC_FASTSIM_ONNX
+  if (o2::zdc::ZDCSimParam::Instance().useZDCFastSim && mFastSimModel != nullptr) {
```
and where do you invoke the model ? Is this not done yet?

SwirtaB replied:
It is done at line 2446.

SwirtaB replied:
mFastSimResults.push_back(mFastSimModel->getChannels(particle));

sawenzel replied:
Ok I see. Thanks.

sawenzel commented on Detectors/ZDC/simulation/src/Detector.cxx:
```diff
@@ -2398,6 +2430,22 @@ void Detector::BeginPrimary()
+
+#ifdef ZDC_FASTSIM_ONNX
+  if (o2::zdc::ZDCSimParam::Instance().useZDCFastSim && mFastSimModel != nullptr) {
+    std::vector<float> particle = {static_cast<float>(mCurrentPrincipalParticle.Energy()),
```
So this looks like you are creating a new std::vector at each invocation which is overhead.
Consider using a different fixed-size type or struct that can be passed to the fast sim model, so that it can be constructed on the stack.

SwirtaB replied:
std::vector still will be created at each invocation inside fast sim model wrapper. What I think can do, is to change interface so the setData method will get mCurrentPrincipalParticle and extract necessery data itself. What do you think about it?

I think I will roll back to std::array for now and investigate more generic solution after gathering more information.

SwirtaB replied:
Fixed in commit [3e2161d](https://github.com/AliceO2Group/AliceO2/pull/8396/commits/3e2161d5bc10386f37db67ce3d2f8ad15cb1cea1)

sawenzel commented on Detectors/ZDC/simulation/CMakeLists.txt:
```diff
@@ -9,13 +9,23 @@
+                PUBLIC_LINK_LIBRARIES O2::SimulationDataFormat O2::ZDCBase
+                                        O2::DataFormatsZDC O2::CCDB O2::SimConfig
+                                        O2::DetectorsRaw O2::Headers)
+endif()
```
Here I think we can avoid code duplication by constructing the list of sources or list of link libraries before `o2_add_library` (with conditional list append). Or use some kind of CMake generator expression.

SwirtaB replied:
List is probably a good idea.

SwirtaB replied:
Fixed in commit [490480f](https://github.com/AliceO2Group/AliceO2/pull/8396/commits/490480f565d3cb2ab5610f1e7d3b8c67801f2b6a)

SwirtaB committed:
Fixed unnecessary vector creation

SwirtaB committed:
Fixed code duplication

SwirtaB commented:
@sawenzel could you please post why fullCI faild? Also if I rember correctly fullCI and o2-cs8 passed on friday. It's strange that it fails now because I haven't made any changes since then.

sawenzel commented on Detectors/ZDC/simulation/src/Detector.cxx:
```diff
@@ -2387,6 +2404,21 @@ void Detector::FinishPrimary()
+    std::fstream output("o2sim-FastSimResult", output.out | output.app);
+    if (!output.is_open()) {
+      LOG(error) << "Could not open file.";
+    }
```
ideally we shouldn't open this at every primary but leave the file handle open until end of simulation

sawenzel commented:
@SwirtaB : The errors in o2-dataflow are genuine. This is one example where FastSim is disabled so you need to add more protection.

sawenzel reviewed (changes requested):
Still some minor build issues to solve. Otherwise it looks good.

SwirtaB committed:
Updated interface

Interface is now more generic and wil be easier
to maintain in future
Added additional checks in CMake to enforce conditional building

SwirtaB committed:
Revert to open file on save

Had to revert due to strange runtime error.
Error was caused by including <fstream> in Detecot.h.

SwirtaB commented:
I have encountered strange runtime errors so whole O2 rebuild was required. That is why I have done push force.

Some summary:
 - I have greatly modified interface so that it is more generic and will be much easier to maintain in future.
 - According to my internal tests current implementation is on par with previous or slightly faster (due to droping unnecessery copies).
 - I had to roll back to open file on each FinishPrimary due to strange runtime error caused by including fstream in Detector.h. Some other solution is required.

SwirtaB committed:
Fixed

copyright notice
o2checkcode errors

SwirtaB committed:
Added additional guards

SwirtaB commented:
Looks like dataflow is building correctly. Any idea why macOS fails?

sawenzel commented:
I think this is due to the changes that you did in your latest force-push to the models. In general, it's not a good idea to mix several new features into one PR. It would have been better to just work on the integration first.

SwirtaB committed:
Added virtual destructors

SwirtaB committed:
Fixed o2checkcode

SwirtaB commented:
All checks are green. Is it good for merge?
